File: companies/Bloomberg/147.insertion-sort-list.py
"""
Approach: Build a sorted list by inserting each original node into its correct place.
Data structure: a dummy head simplifies insertion at the front of the sorted prefix.
Interview logic: before processing each node, dummy.next is sorted. Splicing the next node after the last smaller sorted node preserves that invariant.
Complexity: O(n^2) time, O(1) extra space.
Tests and edge cases: empty list; already sorted list; reverse-sorted list; duplicates stay valid.
"""
from __future__ import annotations
from typing import Optional

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def insertionSortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        cur = head
        while cur:
            nxt = cur.next
            prev = dummy
            while prev.next and prev.next.val < cur.val:
                prev = prev.next
            cur.next = prev.next
            prev.next = cur
            cur = nxt
        return dummy.next
# @lc code=end

File: companies/Bloomberg/147.insertion-sort-list.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

// Definition for singly-linked list.
// struct ListNode {
//     int val;
//     ListNode *next;
//     ListNode() : val(0), next(nullptr) {}
//     ListNode(int x) : val(x), next(nullptr) {}
//     ListNode(int x, ListNode *next) : val(x), next(next) {}
// };


class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        /*
        Approach: build a sorted list behind a dummy head. For each original
        node, find the insertion point in the sorted prefix and splice the node
        there.

        Complexity: O(n^2) time in the worst case, O(1) extra space.
        */
        ListNode dummy(0);
        ListNode* cur = head;
        while (cur) {
            ListNode* next = cur->next;
            ListNode* prev = &dummy;
            while (prev->next && prev->next->val < cur->val) prev = prev->next;
            cur->next = prev->next;
            prev->next = cur;
            cur = next;
        }
        return dummy.next;
    }
};
